exploratory_data_analysis/pubtator_exporatory_data_analysis.py

```diff
@@ -0,0 +1,257 @@
+        writer = csv.DictWriter(
+            outfile,
+            fieldnames=[
+                "batch_folder",
+                "doc_id",
+                "doi",
+                "pmc",
+                "pmid",
+                "section",
+                "published_year",
+            ],
+            delimiter="\t",
+        )
+        writer.writeheader()
+
+        # Cycle through each batch file
+        for batch_directory in pubtator_abstract_batch:
+
+            # Cycle through each document
+            for doc_obj in tqdm.tqdm(PubtatorTarIterator(batch_directory)):
+                doc_id = doc_obj.xpath("id/text()")
+                doi = doc_obj.xpath("passage/infon[@key='article-id_doi']/text()")
+                pmc_id = doc_obj.xpath("passage/infon[@key='article-id_pmc']/text()")
+                pmid = doc_obj.xpath("passage/infon[@key='article-id_pmid']/text()")
+                sections = doc_obj.xpath("passage/infon[@key='section_type']/text()")
+                section_type = doc_obj.xpath("passage/infon[@key='type']/text()")
+                year = doc_obj.xpath("passage/infon[@key='year']/text()")
+
+                section_headers = "|".join(sorted(list(set(sections))))
+                section_headers_type = "|".join(
+                    sorted(list(set(map(lambda x: x.upper(), section_type))))
+                )
+
+                writer.writerow(
+                    {
+                        "batch_folder": batch_directory.name,
+                        "doc_id": doc_id[0],
+                        "doi": doi[0] if len(doi) > 0 else "",
+                        "pmc": pmc_id[0] if len(pmc_id) > 0 else "",
+                        "pmid": pmid[0] if len(pmid) > 0 else "",
+                        "section": section_headers
+                        if len(section_headers) > 0
+                        else section_headers_type,
+                        "published_year": year[0]
+                        if len(year) > 0
+                        else 0,  # Shouldnt get to the else statement
+                    }
+                )
+
+
+# # Analyze Abstract/Full Text Dataset
+
+pubtator_central_metadata_df = pd.read_csv("output/pmc_metadata.tsv.xz", sep="\t")
+print(pubtator_central_metadata_df.shape)
+pubtator_central_metadata_df.head()
+
+# ## Sanity Check the data
+
+# Sanity check that all documents have a published year
+(pubtator_central_metadata_df.query("published_year.isnull()").shape)
+
+# Do all documents have an id?
+(pubtator_central_metadata_df.query("doc_id.isnull()").shape)
+
+# Do all documents have a pmid?
+(pubtator_central_metadata_df.query("pmid.isnull()").shape)
+
+# Do all documents have a pmc id?
+(pubtator_central_metadata_df.query("pmc.isnull()").shape)
+
+# Do all documents have a doi?
+(pubtator_central_metadata_df.query("doi.isnull()").shape)
+
+# ## Published Year Distribution
+
+(pubtator_central_metadata_df.sort_values("published_year").published_year.unique())
```

Comment:
What's year = 0?

Reply:
Does that mean the year isn't available?

Reply:
Yup. I get annoyed when pandas type cast NA as a float (NA = numpy's nan value), so I decided to fill in documents that don't have a year with a place holder. 0 made sense in this case as I'm 100% sure these documents haven't be around since the beginning of time 😂.